connman/src/iptables.c

```diff
@@ -2282,28 +2285,45 @@ static int iptables_replace(struct connman_iptables *table,
 					struct iptables_replace *r)
 {
 	int err;
+	int level;
+	int optname;
+	socklen_t optlen;
 
 	switch (r->type) {
 	case AF_INET:
 		if (!r->r)
 			return -EINVAL;
 
-		err = setsockopt(table->ipt_sock, IPPROTO_IP,
-				IPT_SO_SET_REPLACE, r->r,
-				sizeof(*r->r) + r->r->size);
+		level = IPPROTO_IP;
+		optname = IPT_SO_SET_REPLACE;
+		optlen = sizeof(*r->r) + r->r->size;
 		break;
 	case AF_INET6:
 		if (!r->r6)
 			return -EINVAL;
 
-		err = setsockopt(table->ipt_sock, IPPROTO_IPV6,
-				IP6T_SO_SET_REPLACE, r->r6,
-				sizeof(*r->r6) + r->r6->size);
+		level = IPPROTO_IPV6;
+		optname = IP6T_SO_SET_REPLACE;
+		optlen = sizeof(*r->r6) + r->r6->size;
 		break;
 	default:
 		return -EINVAL;
 	}
 
+	err = util_lock_file(&table->lock_fd, XT_LOCK_NAME);
+	if (err) {
+		DBG("Getting iptables lock failed.");
+		if (err == -ENOLCK)
+			DBG("Timeout not implemented - fail miserably");
+
+		return -EINVAL;
+	}
+
+	err = setsockopt(table->ipt_sock, level, optname,
+				(level == IPPROTO_IP ? r->r : r->r6), optlen);
```

Comment:
TODO do this differently to avoid warning at build.

Reply:
Either casting into `void *ptr` or doing it later as `(void*)r->r` ... Not sure which is the cleaner, but this avoids having to call lock separately.

Reply:
a748a42d6e2a311be26589a251adf90374c59125 might be cleaner.